gk/fib.c

```diff
@@ -1664,3 +1666,329 @@ del_fib_entry(const char *ip_prefix, struct gk_config *gk_conf)
+{
+	struct gk_fib_ether_dump_entry *eth_tbl = rte_calloc(
+		NULL, 1, sizeof(struct gk_fib_ether_dump_entry), 0);
+	if (eth_tbl == NULL)
+		return -1;
+
+	fillup_gk_fib_ether_dump_entry(fib->u.gateway.eth_cache, eth_tbl);
+
+	dentry->num_ether_entries = 1;
+	dentry->ether_tbl = eth_tbl;
+
+	return 0;
+}
+
+static int
+fillup_neighbor_fib_dump_entry(
+	struct gk_fib_dump_entry *dentry, struct gk_fib *fib)
+{
+	int num_entries = 0;
+	uint32_t next = 0;
+	int32_t index;
+	const void *key;
+	void *data;
+	struct neighbor_hash_table *neigh_ht;
+	struct gk_fib_ether_dump_entry *eth_tbl;
+	struct gk_fib_ether_dump_entry *entry;
+
+	if (dentry->addr.proto == ETHER_TYPE_IPv4)
+		neigh_ht = &fib->u.neigh;
+	else
+		neigh_ht = &fib->u.neigh6;
+
+	eth_tbl = rte_calloc(NULL, neigh_ht->tbl_size,
+		sizeof(struct gk_fib_ether_dump_entry), 0);
+	if (eth_tbl == NULL)
+		return -1;
+
+	index = rte_hash_iterate(neigh_ht->hash_table,
+		(void *)&key, &data, &next);
+	while (index >= 0) {
+		struct ether_cache *eth_cache = &neigh_ht->cache_tbl[index];
+		entry = &eth_tbl[num_entries++];
+
+		fillup_gk_fib_ether_dump_entry(eth_cache, entry);
+
+		index = rte_hash_iterate(neigh_ht->hash_table,
+			(void *)&key, &data, &next);
+	}
+
+	dentry->num_ether_entries = num_entries;
+	dentry->ether_tbl = eth_tbl;
+
+	return 0;
+}
+
+static int
+fillup_gk_fib_dump_entry(struct gk_fib_dump_entry *dentry, struct gk_fib *fib)
+{
+	int ret = -1;
+	dentry->action = fib->action;
+	switch (fib->action) {
+	case GK_FWD_GRANTOR:
+		ret = fillup_grantor_fib_dump_entry(dentry, fib);
+		break;
+
+	case GK_FWD_GATEWAY_FRONT_NET:
+		/* FALLTHROUGH */
+	case GK_FWD_GATEWAY_BACK_NET:
+		ret = fillup_gateway_fib_dump_entry(dentry, fib);
+		break;
+
+	case GK_FWD_NEIGHBOR_FRONT_NET:
+		/* FALLTHROUGH */
+	case GK_FWD_NEIGHBOR_BACK_NET:
+		ret = fillup_neighbor_fib_dump_entry(dentry, fib);
+		break;
+
+	case GK_DROP:
+		ret = 0;
+		break;
+
+	default:
+		rte_panic("Invalid FIB action (%u) at %s with lcore %u\n",
+			fib->action, __func__, rte_lcore_id());
+		break;
+	}
+
+	return ret;
+}
+
+static int
+list_ipv4_fib_entries(lua_State *l, struct gk_lpm *ltbl,
+	char *handler, void *arg)
+{
+	int index;
+	struct gk_fib *fib;
+	const struct rte_lpm_rule *re4;
+	struct rte_lpm_iterator_state state;
+	struct gk_fib_dump_entry dentry;
+	int ret = rte_lpm_iterator_state_init(ltbl->lpm, 0, 0, &state);
+	if (ret < 0) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"gk: failed to initialize the lpm rule iterator state at %s!\n",
+			__func__);
+		goto out;
+	}
+
+	index = rte_lpm_rule_iterate(&state, &re4);
+	while (index >= 0) {
+		memset(&dentry, 0, sizeof(dentry));
+		dentry.addr.proto = ETHER_TYPE_IPv4;
+		dentry.addr.ip.v4.s_addr = htonl(re4->ip);
+		dentry.len = state.depth;
+		fib = &ltbl->fib_tbl[re4->next_hop];
+		ret = fillup_gk_fib_dump_entry(&dentry, fib);
+		if (ret < 0)
+			goto out;
+
+		lua_getglobal(l, handler);
+		lua_pushlightuserdata(l, &dentry);
```

Comment:
If you replace `lua_pushlightuserdata()` with `luaL_pushcdata()`, we don't need `get_fib_ether_dump_entry_mac()` and its siblings.

Reply:
When I tried to call ``luaL_get_ctypeid()`` on ``struct gk_config *`` and ``struct gk_fib_dump_entry *``, the return values are the same, is this expected?

Reply:
Yes, it's the same Lua state. I will push the latest code although it hasn't incorporated all the suggestions.

Reply:
Open an issue on GitHub to address this. Mark it as being at the advanced milestone.